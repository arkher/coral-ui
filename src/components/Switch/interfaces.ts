export type SwitchProps = {
  onChange(value: boolean): void;
  value: boolean;
  disabled?: boolean;
};
