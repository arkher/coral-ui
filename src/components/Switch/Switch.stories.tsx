import React from 'react';
import { withDesign } from 'storybook-addon-designs';
import Switch from './Switch';

export default {
  title: 'Componente/Switch',
  component: Switch,
  argTypes: {
    label: {
      control: { type: 'text' },
      defaultValue: 'Label',
    },
    labelStyle: {
      control: { type: 'none' },
    },
    required: {
      control: { type: 'none' },
    },
    disabled: {
      control: { type: 'none' },
    },
    onChange: {
      control: { type: 'none' },
    },
  },
  decorators: [withDesign],
};

const urlHandoff =
  'https://www.figma.com/file/3raVfIADTUZCzFOOaQ9PMQ/HANDOFF-%7C-Core-Components-Mobile-%7C-Institucional?node-id=209%3A3382';

// Stories
export const Unchecked = (): React.ReactNode => (
  <Switch onChange={() => undefined} value={false} />
);

Unchecked.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const UncheckedDisabled = (): React.ReactNode => (
  <Switch onChange={() => undefined} value={false} disabled />
);

UncheckedDisabled.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Checked = (): React.ReactNode => (
  <Switch onChange={() => undefined} value />
);

Checked.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const CheckedDisabled = (): React.ReactNode => (
  <Switch onChange={() => undefined} value disabled />
);

CheckedDisabled.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};
