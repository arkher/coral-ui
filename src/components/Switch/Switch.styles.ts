import { useTheme } from '@shopify/restyle';
import { StyleSheet, ViewStyle } from 'react-native';
import { borderWidthValues, Theme } from '../../themes';

type UseStyleProps = {
  checked: boolean;
  disabled?: boolean;
};

interface Styles {
  circleStyle: ViewStyle;
  containerStyle: ViewStyle;
}

const useStyles = ({
  checked,
  disabled,
}: UseStyleProps): StyleSheet.NamedStyles<Styles> => {
  const theme = useTheme<Theme>();

  let borderColor = theme.colors['neutral-base'];
  let circleColor = theme.colors['neutral-dark'];
  let backgroundColor = theme.colors.white;

  if (!checked) {
    borderColor = theme.colors['neutral-base'];
    circleColor = theme.colors['neutral-dark'];
    backgroundColor = theme.colors.white;
    if (disabled) {
      borderColor = theme.colors['neutral-light'];
      circleColor = theme.colors['neutral-light'];
    }
  } else {
    circleColor = theme.colors.white;
    backgroundColor = theme.colors['primary-base'];
    if (disabled) {
      backgroundColor = theme.colors['neutral-light'];
    }
  }
  return StyleSheet.create<Styles>({
    circleStyle: {
      width: 16,
      height: 16,
      borderRadius: theme.borderRadii.circular,
      backgroundColor: circleColor,
    },
    containerStyle: {
      width: 40,
      height: 24,
      padding: theme.spacing.quark,
      borderRadius: 25,
      borderWidth: checked ? borderWidthValues.none : borderWidthValues.sm,
      borderColor,
      backgroundColor,
      justifyContent: 'center',
    },
  });
};

export default useStyles;
