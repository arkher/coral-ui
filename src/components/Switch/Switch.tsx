import React, { useEffect, useState } from 'react';
import { Pressable } from 'react-native';
import Animated from 'react-native-reanimated';
import { SwitchProps } from './interfaces';
import useStyles from './Switch.styles';

const DSSwitch: React.FC<SwitchProps> = ({
  onChange,
  value,
  disabled,
}: SwitchProps) => {
  const [switchTranslate] = useState(new Animated.Value(value ? 16 : 0));
  const styles = useStyles({ checked: value, disabled });
  useEffect(() => {
    if (value) {
      Animated.spring(switchTranslate, {
        toValue: 16,
        mass: 1,
        damping: 16,
        stiffness: 120,
        overshootClamping: false,
        restSpeedThreshold: 0.001,
        restDisplacementThreshold: 0.001,
      }).start();
    } else {
      Animated.spring(switchTranslate, {
        toValue: 0,
        mass: 1,
        damping: 16,
        stiffness: 120,
        overshootClamping: false,
        restSpeedThreshold: 0.001,
        restDisplacementThreshold: 0.001,
      }).start();
    }
  }, [value, switchTranslate]);

  const memoizedOnSwitchPressCallback = React.useCallback(() => {
    onChange(!value);
  }, [value, onChange]);

  return (
    <Pressable onPress={memoizedOnSwitchPressCallback} testID="ds-switch">
      <Animated.View style={styles.containerStyle}>
        <Animated.View
          style={[
            styles.circleStyle,
            {
              transform: [
                {
                  translateX: switchTranslate,
                },
              ],
            },
          ]}
        />
      </Animated.View>
    </Pressable>
  );
};

export default DSSwitch;
