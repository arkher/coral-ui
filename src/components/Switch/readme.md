# switch

## Properties

| Property | Attribute | Description | Type     | Default            |
| -------- | --------- | ----------- | -------- | ------------------ |
| `label`  | `label`   |             | `string` | `'Button primary'` |


## Events

| Event     | Description | Type               |
| --------- | ----------- | ------------------ |
| `onClick` |             | `CustomEvent<any>` |


----------------------------------------------

Design System - 4'Devs ðŸš€
