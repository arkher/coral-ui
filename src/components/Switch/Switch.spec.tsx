/* eslint-disable global-require */
import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import Switch from './Switch';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

it('should be marked', () => {
  let value = false;
  const { getByTestId } = render(
    <ThemeProvider theme={themeMaestro}>
      <Switch
        value={value}
        onChange={() => {
          value = true;
        }}
      />
    </ThemeProvider>,
  );

  fireEvent.press(getByTestId('ds-switch'));
  expect(value).toBe(true);
});

test('should execute onchange function', async () => {
  const onEventMock = jest.fn();

  const { getByTestId } = render(
    <ThemeProvider theme={themeMaestro}>
      <Switch value onChange={onEventMock} />
    </ThemeProvider>,
  );

  fireEvent.press(getByTestId('ds-switch'));
  expect(onEventMock).toHaveBeenCalledTimes(1);
});

test('should Checkbox render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <Switch value onChange={() => undefined} />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
