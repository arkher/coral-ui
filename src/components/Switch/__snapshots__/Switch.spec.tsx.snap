// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Checkbox render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="ds-switch"
>
  <View
    style={
      Object {
        "backgroundColor": "#0A7E19",
        "borderColor": "#B4BBC2",
        "borderRadius": 25,
        "borderWidth": 0,
        "height": 24,
        "justifyContent": "center",
        "padding": 4,
        "width": 40,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 9999,
            "height": 16,
            "width": 16,
          },
          Object {
            "transform": Array [
              Object {
                "translateX": AnimatedValue {
                  " __value": 16,
                },
              },
            ],
          },
        ]
      }
    />
  </View>
</View>
`;
